csrc/executor_kernel_arg.cpp

```diff
@@ -125,14 +125,16 @@ std::unique_ptr<TensorArgAbstract> getTensorArg(
 } // namespace
 
 KernelArgumentHolder KernelArgumentHolder::createKernelArgumentHolder(
-    const c10::ArrayRef<c10::IValue>& inputs) {
+    const c10::ArrayRef<c10::IValue>& inputs,
+    std::optional<int8_t> selected_device) {
   if (inputs.empty()) {
     // default to device 0
     KernelArgumentHolder args;
-    args.setDeviceIndex(0);
+    args.setDeviceIndex(
+        selected_device.has_value() ? selected_device.value() : (int8_t)0);
     return args;
   }
-  auto device_index = getCommonDeviceCUDA(inputs);
+  auto device_index = getCommonDeviceCUDA(inputs, selected_device);
```

Comment:
So, in this case `inputs` is not empty. Does passing `selected_device` matter?

Reply:
Also, please add an error check here, i.e., `0 <= device_index < max_device_index`

Reply:
> So, in this case `inputs` is not empty. Does passing `selected_device` matter?

Actually, it does matter since even if `inputs` is not empty, it may not contain any device tensors, so in that case the optional device index would be used.

Reply:
Sorry you're right I am missing some comments explaining this. The parameter is used if given:
1. If inputs contains tensors their devices must match `selected_device` if it is provided.
2. Otherwise, `selected_device` will be used if provided, and if not, will default to device 0.
The logic behind (1) is that selected_device is used in cases like `torch.full(..., device='cuda:1')` which should not quietly change to running on device 0 if `cuda:0` devices are provided; in PyTorch that would be a mismatched device error, so it is here as well. There is always the option to omit the parameter to default to device 0. I will update the doxygen for all changed interfaces.